Source/WebCore/loader/cache/CachedResource.cpp

```diff
@@ -916,6 +914,87 @@ void CachedResource::setLoadPriority(const std::optional<ResourceLoadPriority>&
         m_loadPriority = DefaultResourceLoadPriority::forResourceType(type());
 }
 
+CachedResource::ResponseData::ResponseData(CachedResource& resource)
+    : m_decodedDataDeletionTimer(resource, &CachedResource::destroyDecodedData, deadDecodedDataDeletionIntervalForResourceType(resource.type()))
+{
+}
+
+CachedResource::ResponseData& CachedResource::mutableResponseData() const
+{
+    if (!m_response)
+        m_response = makeUnique<ResponseData>(*const_cast<CachedResource*>(this));
+    return *m_response;
+}
+
+ResourceResponse& CachedResource::mutableResponse()
+{
+    return mutableResponseData().m_response;
+}
+
+const ResourceResponse& CachedResource::response() const
+{
+    if (!m_response) {
+        static LazyNeverDestroyed<ResourceResponse> staticEmptyResponse;
+        static std::once_flag onceFlag;
+        std::call_once(onceFlag, [&] {
+            staticEmptyResponse.construct();
+        });
+        return staticEmptyResponse;
+    }
+    return m_response->m_response;
+}
+
+void CachedResource::stopDecodedDataDeletionTimer()
+{
+    if (!m_response)
+        return;
+    mutableResponseData().m_decodedDataDeletionTimer.stop();
+}
+
+void CachedResource::restartDecodedDataDeletionTimer()
+{
+    mutableResponseData().m_decodedDataDeletionTimer.restart();
+}
+
+const ResourceError& CachedResource::resourceError() const
+{
+    if (!m_response) {
+        static LazyNeverDestroyed<ResourceError> emptyError;
+        static std::once_flag onceFlag;
+        std::call_once(onceFlag, [&] {
+            emptyError.construct();
+        });
+        return emptyError;
```

Comment:
I don't think we need this, since the ResourceError lives in the ResponseData, and we can just return the one from response().

Reply:
response() is a ResourceResponse, not ResponseData.